--- tmux.c.orig	2016-03-05 20:55:49.000000000 +0300
+++ tmux.c	2016-05-05 04:09:58.507575080 +0300
@@ -194,11 +194,11 @@
 
 	if (setlocale(LC_CTYPE, "en_US.UTF-8") == NULL) {
 		if (setlocale(LC_CTYPE, "") == NULL)
-			errx(1, "invalid LC_ALL, LC_CTYPE or LANG");
+			warnx("invalid LC_ALL, LC_CTYPE or LANG");
 		s = nl_langinfo(CODESET);
 		if (strcasecmp(s, "UTF-8") != 0 &&
 		    strcasecmp(s, "UTF8") != 0)
-			errx(1, "need UTF-8 locale (LC_CTYPE) but have %s", s);
+			warnx("need UTF-8 locale (LC_CTYPE) but have %s", s);
 	}
 
 	setlocale(LC_TIME, "");
